Feature: PowerVS cluster creation with dynamic resource creation

Karthik-K-N opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:

This PR adds the ability to create required resources for creating Power VS cluster dynamically eliminating user to precreate resources before creating cluster. A reference to proposal has been added which contains the detailed implementation plan for this feature.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add ability to create required resource dynamically for PowerVS cluster creation
```

Proposal: https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1488
API Pr: https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1485
            https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1592

TODO:
* [ ] Fix and add unit test cases
* [ ] Add default and validating webhooks
* [ ] Update cos object creation and deletion logic

Karthik-K-N commented:
/cc @mkumatag @dharaneeshvrd @Amulyam24 @Prajyot-Parab

mkumatag commented:
> /cc @mkumatag @dharaneeshvrd @Amulyam24 @Prajyot-Parab

team lets review this code by this weekend

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	// if DHCP server id is set means the server is already created
+	if s.GetDHCPServerID() != nil {
+		s.Info("DHCP server id is set")
+		err := s.checkDHCPServerStatus()
```
```suggestion
		if err := s.checkDHCPServerStatus(); err != nil
```

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	if err != nil {
+		return err
+	}
+	if dhcpServer == nil {
```
we can remove this check, dhcpclient.get returns either err or dhcpserver obj. If err is nil, dhcpserver obj won't be nil.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	switch resourceType {
+	case ServiceInstance:
+		if s.ServiceInstance() == nil || s.ServiceInstance().Name == nil {
+			return pointer.String(fmt.Sprintf("%s-serviceInstance", s.InfraCluster().GetName()))
```
Regarding the suffix, sometime back had a discussion with IPI team and decided to use common suffix, for powervs and other few resources which hypershift infra creation also follows currently. Also keeping longer suffix would make difficult to find the cluster resources in ibm cloud portal. Keeping it short would be better IMO.
powervs - pvs
vpc - vpc
vpc subnet - vpc-sn
transit gateway - tg
cos - image-registry - this alone shared by ibmcloud platform as well
Please see whether using this would make sense!

Karthik-K-N replied:
We will handle this in next phase.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		PortMin:   core.Int64Ptr(int64(s.APIServerPort())),
+		PortMax:   core.Int64Ptr(int64(s.APIServerPort())),
+	})
+	_, _, err = s.IBMVPCClient.CreateSecurityGroupRule(options)
```
```suggestion
	if _, _, err = s.IBMVPCClient.CreateSecurityGroupRule(options); err != nil {
```

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	//TODO: Should we handle a case where there are no subnet specified
+	for _, subnet := range s.IBMPowerVSCluster.Spec.VPCSubnets {
+		if subnet.Name == nil {
+			subnet.Name = pointer.String(fmt.Sprintf("%s-vpcsubnet", s.InfraCluster().GetName()))
```
No need to keep this in `GetServiceName()`?

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	if err != nil {
+		return "", err
+	}
+	if vpcSubnet == nil {
```
Not required like dhcp get

Karthik-K-N replied:
I think its required as vpcsubnet may not be found in cloud

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		s.Error(err, "failed to get vpc")
+		return "", err
+	}
+	if vpcDetails == nil {
```
Not required like dhcp get. I have checked, you can reverify once in vpc sdk get call.

Karthik-K-N replied:
I think its required as vpc may not be found in cloud

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	}
+
+	apiKey := os.Getenv("IBMCLOUD_API_KEY")
+	if apiKey == "" {
+		fmt.Printf("ibmcloud api key is not provided, set %s environmental variable", "IBMCLOUD_API_KEY")
+	}
+	apiKey = APIKEY
```
how APIKEY is set?
I m seeing an To-Do remove for this. You may need to properly handle here!

Karthik-K-N replied:
Its handled in latest changes by Prajyot

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		return nil, fmt.Errorf("error getting id for resource group %s, %w", *s.ResourceGroup(), err)
+	}
+
+	//TODO(karthik-k-n)(Doubt): should this be fetched or hardcoded
+	//_, servicePlanID, err := s.CatalogClient.GetServiceInfo(powerVSService, powerVSServicePlan)
+	//if err != nil {
+	//	return nil, fmt.Errorf("error retrieving id info for powervs service %w", err)
+	//}
+
```
IMO can be hardcoded! It is not going to change!

dharaneeshvrd commented on controllers/ibmpowervscluster_controller.go:
```diff
@@ -101,25 +107,168 @@ func (r *IBMPowerVSClusterReconciler) Reconcile(ctx context.Context, req ctrl.Re
+	if err := clusterScope.ReconcileCOSInstance(); err != nil {
+		conditions.MarkFalse(powerVSCluster, infrav1beta2.COSInstanceReadyCondition, infrav1beta2.COSInstanceReconciliationFailedReason, capiv1beta1.ConditionSeverityError, err.Error())
+		return reconcile.Result{}, err
+	}
```
don't infrav1beta2.COSInstanceReadyCondition set to be true here?

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		return fmt.Errorf("error listing transit gateway connections: %w", err)
+	}
+
+	for _, conn := range tgConnections.Connections {
```
Handled the case when connections are 0 but added a TODO to verify.

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			s.Error(err, "error creating vpc subnet")
+			return err
+		}
+		if subnetID != nil {
```
I think its required as the subnet may not be exist in status

Karthik-K-N replied:
You are correct its redundant check, I removed it.

Karthik-K-N commented:
Thanks for Dhraneesh for reviewing, I have addressed the comments and marked it as resolved whichever is possible, We will discuss for unresolved comments to finalize the approaches.

Amulyam24 commented on cloud/scope/powervs_machine.go:
```diff
@@ -141,25 +153,38 @@ func NewPowerVSMachineScope(params PowerVSMachineScopeParams) (scope *PowerVSMac
+	if serviceInstance == nil {
+		return nil, fmt.Errorf("service instance %s is not yet created", serviceInstanceName)
+	}
+	if *serviceInstance.State != "active" {
```
```suggestion
	if *serviceInstance.State != string(infrav1beta2.ServiceInstanceStateActive) {
```

Amulyam24 commented on cloud/scope/powervs_machine.go:
```diff
@@ -295,24 +485,89 @@ func (m *PowerVSMachineScope) DeleteMachine() error {
+		m.Info("cos service instance is nil")
+		return nil, err
+	}
+	if *serviceInstance.State != "active" {
```
```suggestion
	if *serviceInstance.State != string(infrav1beta2.ServiceInstanceStateActive) {
```

Amulyam24 commented on pkg/cloud/services/resourcecontroller/service.go:
```diff
@@ -43,14 +70,114 @@ func (s *Service) GetResourceInstance(getResourceInstanceOptions *resourcecontro
+	case 1:
+		return &serviceInstancesList[0], nil
+	default:
+		errStr := fmt.Errorf("there exist more than one service instance ID with with same name %s, Try setting serviceInstance.ID", name)
```
```suggestion
		errStr := fmt.Errorf("there exist more than one service instance ID with same name %s, Try setting serviceInstance.ID", name)
```

Amulyam24 commented on pkg/cloud/services/resourcecontroller/service.go:
```diff
@@ -43,14 +70,114 @@ func (s *Service) GetResourceInstance(getResourceInstanceOptions *resourcecontro
+	case 1:
+		return &serviceInstancesList[0], nil
+	default:
+		errStr := fmt.Errorf("there exist more than one COS instance ID with with same name %s, Try setting serviceInstance.ID", name)
```
```suggestion
		errStr := fmt.Errorf("there exist more than one COS instance ID with same name %s, Try setting serviceInstance.ID", name)
```

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +292,1545 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// getResourceGroupID retrieving id of resource group.
+func (s *PowerVSClusterScope) getResourceGroupID() (string, error) {
```
made changes to set resource group id in status.

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+
+func (s *PowerVSClusterScope) createCOSBucket() error {
+	input := &s3.CreateBucketInput{
+		Bucket: pointer.String(s.COSInstance().BucketName),
```
There's a scenario where when we retry bucket creation immediately with the same name,  it throws an error saying bucket exists even though the service instance and bucket get deleted successfully. Maybe we can handle by adding a random string to the bucket name?
(check out `util.RandomString()` from `sigs.k8s.io/cluster-api/util`)

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// deleteResource returns true or false to decide on deleting provided resource.
+func (s *PowerVSClusterScope) deleteResource(resourceType ResourceType) bool { //nolint:gocyclo
```
or some other meaning func name. 
```suggestion
func (s *PowerVSClusterScope) resourceCreatedByController(resourceType ResourceType) bool { //nolint:gocyclo
```
same as other func, the name looks like, its deleting some resource

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			}
+
+			if lb != nil && lb.ProvisioningStatus != nil && *lb.ProvisioningStatus != string(infrav1beta2.VPCLoadBalancerStateDeletePending) {
+				_, err = s.IBMVPCClient.DeleteLoadBalancer(&vpcv1.DeleteLoadBalancerOptions{
+					ID: lb.ID,
+				})
+				if err != nil {
```
```suggestion
				if _, err = s.IBMVPCClient.DeleteLoadBalancer(&vpcv1.DeleteLoadBalancerOptions{
					ID: lb.ID,
				}); err != nil {
```

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return fmt.Errorf("error listing transit gateway connections: %w", err)
+		}
+
+		if tgConnections.Connections != nil && len(tgConnections.Connections) > 0 {
```
this if is not required since Connections is not a pointer to the array, if the connection len is zero, it won't loop.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			}
+		}
+
+		_, err = s.TransitGatewayClient.DeleteTransitGateway(&tgapiv1.DeleteTransitGatewayOptions{
+			ID: s.IBMPowerVSCluster.Status.TransitGateway.ID,
+		})
+
+		if err != nil {
```
```suggestion
		if _, err = s.TransitGatewayClient.DeleteTransitGateway(&tgapiv1.DeleteTransitGatewayOptions{
			ID: s.IBMPowerVSCluster.Status.TransitGateway.ID,
		}); err != nil {
```

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return fmt.Errorf("error fetching DHCP server: %w", err)
+		}
+
+		err = s.IBMPowerVSClient.DeleteDHCPServer(*server.ID)
+		if err != nil {
+			return fmt.Errorf("error deleting the DHCP server: %w", err)
+		}
```
```suggestion
		if err = s.IBMPowerVSClient.DeleteDHCPServer(*server.ID); err != nil {
			return fmt.Errorf("error deleting the DHCP server: %w", err)
		}
```

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return fmt.Errorf("error fetching the subnet: %w", err)
+		}
+
+		_, err = s.IBMVPCClient.DeleteSubnet(&vpcv1.DeleteSubnetOptions{
+			ID: net.ID,
+		})
+		if err != nil {
```
inline if ?

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			}
+			return fmt.Errorf("error fetching the subnet: %w", err)
+		}
+
```
I think you need to wait for load balancer to get deleted before proceeding to subnet deletion or else it will fail.
Please see whether we need a check here like how a check is there for dhcpserver before deleting powervs instance.

Amulyam24 replied:
The idea behind this was that on `reconcileDelete`, we are not waiting for one resource to get deleted before proceeding to the other resource but rather call delete API on all the resources and return an aggregate of errors at the end. One thing is it logs the error before a successful deletion in the subsequent reconciliations.

I think it might get redundant adding check for LB here again since we check it with `DeleteLoadBalancer` in the previous step.

> Please see whether we need a check here like how a check is there for dhcpserver before deleting powervs instance.

For deleting powers instance, the check was added because even though the delete API returns with success, the service instance is not getting deleted when checked on UI. We thought it would be better to wait and delete any resources within it before deleting the workspace.

dharaneeshvrd replied:
Got it, only concerns me is that the intermediate error status user would see in ibmpowervscluster obj during destroy.
One example could be in future if some other external program query for this obj on an interval, not sure it would be good to throw these intermediate errors.

Amulyam24 replied:
@dharaneeshvrd, will be handling it in next iteration.

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return fmt.Errorf("error fetching the VPC: %w", err)
+		}
+
+		_, err = s.IBMVPCClient.DeleteVPC(&vpcv1.DeleteVPCOptions{
+			ID: vpc.ID,
+		})
+		if err != nil {
```
inline if?

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return fmt.Errorf("cannot delete service instance as network is not yet deleted")
+		}
+
+		_, err = s.ResourceClient.DeleteResourceInstance(&resourcecontrollerv2.DeleteResourceInstanceOptions{
```
inline if

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return nil
+		}
+
+		_, err = s.ResourceClient.DeleteResourceInstance(&resourcecontrollerv2.DeleteResourceInstanceOptions{
```
inline if

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			}
+			return fmt.Errorf("error fetching the VPC: %w", err)
+		}
+
```
Replied via https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1608#discussion_r1494240913

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		return fmt.Errorf("error listing transit gateway connections: %w", err)
+	}
+
+	if len(tgConnections.Connections) == 0 {
```
I have updated the logic, Please take a look

dharaneeshvrd commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1557 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			powerVSAttached = true
+		}
+	}
+	if !powerVSAttached || !vpcAttached {
```
It seems it won't reach here if any one of the case is false since you are already returning from respective case.

Karthik-K-N replied:
yeah, The idea is given both the cases are true, we want to make sure both vpc and powervs is attached in tg connection

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+
+// Zone returns the cluster zone.
+func (s *PowerVSClusterScope) Zone() *string {
+	return s.IBMPowerVSCluster.Spec.Zone
```
what happens when someone calls when this zone is null?!

Karthik-K-N replied:
For zone and resource group we are adding webhook to make it mandatory for user to set when using create infra workflow which is identified with help of annotations. 
But even if we want to be safe then we need to add nil check under each caller of this function.

mkumatag replied:
its always safe to have null check :)

Karthik-K-N replied:
Added check where ever the value is being referred.

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// ResourceGroup returns the cluster resource group.
+func (s *PowerVSClusterScope) ResourceGroup() *infrav1beta2.IBMPowerVSResourceReference {
```
again this * may create issues when something is wrong, wondering if we can have a simple null check?

Karthik-K-N replied:
Added check where ever the value is being referred.

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	if s.Cluster.Spec.ClusterNetwork != nil && s.Cluster.Spec.ClusterNetwork.APIServerPort != nil {
+		return *s.Cluster.Spec.ClusterNetwork.APIServerPort
+	}
+	return 6443
```
can we create a constant somewhere for this port and use it?

Karthik-K-N replied:
Created a constant for it.

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+
+// GetServiceInstanceID get the service instance id.
+func (s *PowerVSClusterScope) GetServiceInstanceID() string {
+	// support the existing worklfow
```
can you add more information here about the flow we written here?

mkumatag replied:
or may be a generic information above which applies for all the Get calls

Karthik-K-N replied:
sure will do

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	if err != nil {
+		return "", err
+	}
+	s.SetStatus(ResourceGroup, infrav1beta2.ResourceReference{ID: &resourceGroupID, ControllerCreated: pointer.Bool(false)})
```
why new setting this value in the Get method? 🤔

Karthik-K-N replied:
We are expecting user to pass resource group name, So we only set the id. The GetResourceGroupID is called from various places but only for the first call the status will be set and in future it will be fetched from status itself.

mkumatag replied:
but its better to call that separately and set the status instead of in this function

Karthik-K-N replied:
Updated.

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+func (s *PowerVSClusterScope) ReconcileServiceInstance() error {
+	serviceInstanceID := s.GetServiceInstanceID()
+	if serviceInstanceID != "" {
+		s.Info("Service instance id is set", "id", serviceInstanceID)
+		serviceInstance, _, err := s.ResourceClient.GetResourceInstance(&resourcecontrollerv2.GetResourceInstanceOptions{
+			ID: &serviceInstanceID,
+		})
+		if err != nil {
+			return err
+		}
+		if serviceInstance == nil {
+			return fmt.Errorf("error failed to get service instance with id %s", serviceInstanceID)
+		}
+		if *serviceInstance.State != string(infrav1beta2.ServiceInstanceStateActive) {
+			return fmt.Errorf("service instance not in active state, current state: %s", *serviceInstance.State)
+		}
+		s.Info("Found service instance and its in active state", "id", serviceInstanceID)
```
I see this can be a code you can run post creation of resource in the end as well.

mkumatag replied:
applies this comment for all other reconciles as well

mkumatag reviewed (changes requested):
first set of review comments

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -59,93 +98,158 @@ type PowerVSClusterScope struct {
+		session:              session,
+		Logger:               params.Logger,
+		Client:               params.Client,
+		patchHelper:          helper,
```
I have moved defer after checking for error https://github.com/Karthik-K-N/cluster-api-provider-ibmcloud/blob/f62f4958cf40d3c2a9aad20c5f03e68e0ba816cd/controllers/ibmpowervscluster_controller.go#L88-L98, I think there is nothing to patch at that moment.

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return pointer.String(fmt.Sprintf("%s-serviceInstance", s.InfraCluster().GetName()))
+		}
+		return s.ServiceInstance().Name
+	case infrav1beta2.ResourceTypeNetwork:
```
Fix this: Currently DHCP server creates network and its of some format as given here, If user provides name this wont work. I think dhcp server name cannot be set as we think. Explore!!

Karthik-K-N replied:
Updated the logic to handle the scenarios when user provided DHCPserver name as well.

mkumatag commented on util/util.go:
```diff
@@ -39,3 +40,160 @@ func GetClusterByName(ctx context.Context, c client.Client, namespace, name stri
+}
+
+// Regions provides a mapping between Power VS and IBM Cloud VPC and IBM COS regions.
+var Regions = map[string]Region{
```
dependent's k8s version is dealt via go.sum file I feel

Karthik-K-N replied:
The issue I saw is, When I tried to use powervs utils it uses k8s version 1.29 but capi is using 1.28, It updated the verision in capi

capi go.mod file
```
        k8s.io/apiextensions-apiserver v0.28.4
-       k8s.io/apimachinery v0.28.4
+       k8s.io/apimachinery v0.29.1
        k8s.io/cli-runtime v0.28.4
        k8s.io/client-go v0.28.4
-       k8s.io/klog/v2 v2.100.1
-       k8s.io/utils v0.0.0-20230406110748-d93618cff8a2
+       k8s.io/klog/v2 v2.110.1
+       k8s.io/utils v0.0.0-20230726121419-3b25d923346b
```

mkumatag commented on api/v1beta2/ibmpowervscluster_types.go:
```diff
@@ -92,6 +97,22 @@ type IBMPowerVSClusterSpec struct {
+// DHCPServer contains the DHCP server configurations
+type DHCPServer struct {
+	// Optional cidr for DHCP private network
+	Cidr *string `json:"cidr,omitempty"`
```
lets add some regex to validate this

Karthik-K-N replied:
Yet to find a valid regex, will update soon

mkumatag commented on api/v1beta2/ibmpowervscluster_types.go:
```diff
@@ -92,6 +97,22 @@ type IBMPowerVSClusterSpec struct {
+	DNSServer *string `json:"dnsServer,omitempty"`
+
+	// Optional name of DHCP Service. Only alphanumeric characters and dashes are allowed (will be prefixed by DHCP identifier)
+	Name *string `json:"name,omitempty"`
```
what if user specifies the id as well?

Karthik-K-N replied:
Added ID field as well.

mkumatag commented on api/v1beta2/ibmpowervscluster_types.go:
```diff
@@ -92,6 +97,22 @@ type IBMPowerVSClusterSpec struct {
+	Name *string `json:"name,omitempty"`
+
+	// Indicates if SNAT will be enabled for DHCP service
+	SnatEnabled *bool `json:"snatEnabled,omitempty"`
```
this should be true by default and you can rename this to `snat` to keep it simple as this is already boolean.

Karthik-K-N replied:
Updated.

Amulyam24 commented:
Review comments from discussion today:

- [ ] 1. Add comments and documentation for easy understanding of code
- [ ] 2. Regex should be handled properly for all resources
- [ ] 3. CIDR regex validation for DHCP server
- [ ] 4. Handle Set() resource in ibmpowervscluster_types.go — complicating the code L243
- [ ] 5. Think about how to test the machine types in staging environment
- [ ] 6. Check why session is being used/added in PowerVSClusterScope L77
- [ ] 7. Why ClusterObject is being used?  just return object name, remove setter
- [ ] 8. Optimise zone handling in PowerVSClusterScope() L125
- [ ] 9. Add webhook validation for VPC region as well L190
- [ ] 10. Rename APIServerPort as DefaultAPIServerPort
- [ ] 11. SetNetworkStatus move to the common Set method
- [ ] 12. Change the order in GetDHCPServerID
- [ ] 13. Remove get for GetDHCPServer in parity with others
- [ ] 14. GetSubnetID() - instead of sending whole object, just send the name 
- [ ] 15. GetSubnetIDs() — no need to return array of string pointers
- [ ] 16. Remove additional check(controllerCreated) in SetVPCSubnetID() L416 L458
- [ ] 17. Check wherever pointers are not necessary(check for nil values)
- [ ] 19. GetLoadBanlancerID() L466 - just pass name 
- [ ] 21. GetLoadBalancerhost — add name to match the value being returned — check for others as well 
- [ ] 22. Move SetCosBucket to common Set method L499
- [ ] 23. Handle GetResourceGroupID()  — don’t call the fetchResourcegroupID()   L522

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -59,93 +74,158 @@ type PowerVSClusterScope struct {
-	if svcEndpoint := endpoints.FetchPVSEndpoint(endpoints.CostructRegionFromZone(*res.RegionID), params.ServiceEndpoint); svcEndpoint != "" {
-		options.IBMPIOptions.URL = svcEndpoint
-		scope.Logger.V(3).Info("Overriding the default powervs service endpoint")
+	if params.IBMPowerVSCluster.Spec.VPC == nil || params.IBMPowerVSCluster.Spec.VPC.Region == nil {
```
Add webhook validation for VPC region as well

Amulyam24 commented on api/v1beta2/types.go:
```diff
@@ -16,6 +16,9 @@ limitations under the License.
 package v1beta2
 
+// APIServerPort is API server port number.
+const APIServerPort int32 = 6443
```
Rename DefaultAPIServerPort

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// SetNetworkStatus set the network status.
+func (s *PowerVSClusterScope) SetNetworkStatus(resource infrav1beta2.ResourceReference) {
```
move to the common Set method

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// GetDHCPServerID returns the DHCP id.
+func (s *PowerVSClusterScope) GetDHCPServerID() *string {
```
Change the order of checking status and spec

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// GetDHCPServer returns the DHCP server details.
+func (s *PowerVSClusterScope) GetDHCPServer() *infrav1beta2.DHCPServer {
```
Rename as DHCPServer

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	if s.IBMPowerVSCluster.Status.VPCSubnet == nil {
+		return nil
+	}
+	subnets := []*string{}
```
maybe no need to return array of string pointers

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		s.IBMPowerVSCluster.Status.VPCSubnet = make(map[string]infrav1beta2.ResourceReference)
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.VPCSubnet[name]; ok {
+		if val.ControllerCreated != nil && *val.ControllerCreated {
```
remove check wherever unecessary

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		s.IBMPowerVSCluster.Status.LoadBalancers = make(map[string]infrav1beta2.VPCLoadBalancerStatus)
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		if val.ControllerCreated != nil && *val.ControllerCreated {
```
remove check wherever unnecessary

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// SetCOSBucketStatus set the COS bucket id.
+func (s *PowerVSClusterScope) SetCOSBucketStatus(cosInstanceID string, controllerCreated bool) {
```
move to common Set method

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	if s.IBMPowerVSCluster.Status.ResourceGroup != nil && s.IBMPowerVSCluster.Status.ResourceGroup.ID != nil {
+		return *s.IBMPowerVSCluster.Status.ResourceGroup.ID, nil
+	}
+	resourceGroupID, err := s.fetchResourceGroupID()
```
maybe handle this in a better way by avoiding calling `fetchResourceGroupID` here

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +237,1610 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// GetLoadBalancerHost will return the hostname of load balancer.
+func (s *PowerVSClusterScope) GetLoadBalancerHost(name string) *string {
```
rename to GetLoadBalancerHostName

Amulyam24 commented on api/v1beta2/ibmpowervscluster_types.go:
```diff
@@ -222,6 +239,14 @@ func (r *IBMPowerVSCluster) SetConditions(conditions capiv1beta1.Conditions) {
+// Set sets the details of the resource.
+func (rf *ResourceReference) Set(resource ResourceReference) {
+	rf.ID = resource.ID
+	if !*rf.ControllerCreated {
```
re-evaluate this check

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -59,93 +74,158 @@ type PowerVSClusterScope struct {
 	logr.Logger
 	Client      client.Client
 	patchHelper *patch.Helper
+	session     *ibmpisession.IBMPISession
```
Reused across multiple function for example https://github.com/Karthik-K-N/cluster-api-provider-ibmcloud/blob/969051ba88de2179b8f4ed9bec8728f0711b6dd4/cloud/scope/powervs_cluster.go#L1434, Need to explore on reusing it even furthur.

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +231,1605 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	case infrav1beta2.ResourceTypeSubnet:
+		return pointer.String(fmt.Sprintf("%s-vpcsubnet", s.InfraCluster()))
+	}
+	return nil
```
handle error if type is not present

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +231,1605 @@ func (s *PowerVSClusterScope) PatchObject() error {
+}
+
+// resourceCreatedByController helps to identify resource created by controller or not.
+func (s *PowerVSClusterScope) resourceCreatedByController(resourceType infrav1beta2.ResourceType) bool { //nolint:gocyclo
```
rename to `isResourceCreatedByController` wherever applicable

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +231,1605 @@ func (s *PowerVSClusterScope) PatchObject() error {
+			return nil, fmt.Errorf("error creating vpc subnet, powervs zone is not set")
+		}
+		region := endpoints.ConstructRegionFromZone(*powerVSZone)
+		vpcZones, err := genUtil.VPCZonesForPowerVSRegion(region)
```
The region mapping is currently from powervs region to vpc https://github.com/Karthik-K-N/cluster-api-provider-ibmcloud/blob/67e8057bad2a4fadf233d4e8636199ba4795b1d1/util/util.go#L77, so we need to fetch powervs region from zone

mkumatag commented on go.mod:
```diff
@@ -12,18 +12,25 @@ replace (
 	github.com/IBM/platform-services-go-sdk v0.59.0
 	github.com/IBM/vpc-go-sdk v0.48.0
+	github.com/blang/semver/v4 v4.0.0
+	github.com/coreos/ignition v0.35.0
```
here is the CVE link -- https://github.com/advisories/GHSA-hj57-j5cw-2mwp, as this is a very old release, I don't think they will even patch this one. Lets keep this schema offline somewhere in the pkg folder.

Prajyot-Parab replied:
fix - https://github.com/Karthik-K-N/cluster-api-provider-ibmcloud/pull/12

mkumatag commented on cloud/scope/powervs_cluster.go:
```diff
@@ -157,3 +246,1660 @@ func (s *PowerVSClusterScope) PatchObject() error {
+
+// fetchResourceGroupID retrieving id of resource group.
+func (s *PowerVSClusterScope) fetchResourceGroupID() (string, error) {
+	if s.ResourceGroup() == nil || s.ResourceGroup().Name == nil {
+		return "", fmt.Errorf("resource group name is not set")
+	}
```
TODO: Use default resource group for the account if not specified.

mkumatag commented on cloud/scope/powervs_machine.go:
```diff
@@ -295,24 +506,100 @@ func (m *PowerVSMachineScope) DeleteMachine() error {
-// GetBootstrapData returns the base64 encoded bootstrap data from the secret in the Machine's bootstrap.dataSecretName.
-func (m *PowerVSMachineScope) GetBootstrapData() (string, error) {
+// DeleteMachineIgnition deletes the ignition associated with machine.
+func (m *PowerVSMachineScope) DeleteMachineIgnition() error {
```
where is this called?

Prajyot-Parab replied:
@Karthik-K-N did you drop it by mistake? I have it in my other PR, if need be I can split it in a diff PR and push.

Karthik-K-N replied:
Updated it in the latest change to delete ignition as part of machine deletion.

mkumatag commented:
@Prajyot-Parab @Karthik-K-N I see linting issue for that file, see if we can somehow ignore that for that entire package

Prajyot-Parab commented:
> @Prajyot-Parab @Karthik-K-N I see linting issue for that file, see if we can somehow ignore that for that entire package

fixed.

mkumatag commented:
I have approved the PR, one of you @Amulyam24 @Prajyot-Parab give lgtm

Prajyot-Parab reviewed (approved):
/lgtm
